crates/bevy_geometry/src/lib.rs

```diff
@@ -0,0 +1,359 @@
+use bevy_math::*;
+use bevy_reflect::Reflect;
+use std::error::Error;
+use std::fmt;
+
+pub trait Primitive3d {
+    /// Returns true if this primitive is entirely on the outside (in the normal direction) of the
+    /// supplied plane.
+    fn outside_plane(&self, plane: Plane) -> bool;
+}
+
+#[derive(Debug, Clone)]
+pub enum PrimitiveError {
+    MinGreaterThanMax,
+    NonPositiveExtents,
+}
+impl Error for PrimitiveError {}
+impl fmt::Display for PrimitiveError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            PrimitiveError::MinGreaterThanMax => {
+                write!(
+                    f,
+                    "AxisAlignedBox minimums must be smaller or equal to the maximums"
+                )
+            }
+            PrimitiveError::NonPositiveExtents => {
+                write!(f, "AxisAlignedBox extents must be greater than zero")
+            }
+        }
+    }
+}
+
+#[derive(Copy, Clone, PartialEq, Debug, Reflect)]
+pub struct Sphere {
+    origin: Vec3,
+    radius: f32,
+}
+
+impl Sphere {
+    /// Get a reference to the sphere's origin.
+    pub fn origin(&self) -> &Vec3 {
```

Comment:
Why are `origin` and `radius` returned by reference? The `Vec3` should hopefully be returned in a register and `f32` is smaller than the reference to it.

Reply:
I was using rust-analyzer to automatically generate setters and getters, but didn't get around to fixing this yet.
I'd like to avoid pub for these to have a more stable interface, it's very possible the underlying data structure will change as we optimize.